File: backend/app/limiters/sliding_counter.py
"""Sliding Window Counter — weighted blend of current + previous fixed window."""

from __future__ import annotations

from ..config import SlidingCounterParams
from .base import RateLimiter


class SlidingCounterLimiter(RateLimiter):
    key = "sliding_counter"

    async def setup(self) -> None:
        self._script = self.load_script("sliding_counter.lua")

    async def evaluate(
        self, client_id: str, params: SlidingCounterParams, now: float
    ) -> tuple[bool, dict, float | None]:
        window_s = params.window_s
        idx = int(now // window_s)
        elapsed = now - idx * window_s
        # Fraction of the previous window still inside the trailing window.
        weight = (window_s - elapsed) / window_s

        res = await self._script(
            keys=[f"{self.key}:{client_id}:{idx}", f"{self.key}:{client_id}:{idx - 1}"],
            args=[params.limit, weight, int(window_s * 1000)],
        )
        allowed = bool(int(res[0]))
        curr, prev = int(res[1]), int(res[2])
        estimate = float(res[3])

        state = {
            "curr_count": curr,
            "prev_count": prev,
            "weight": round(weight, 4),
            "estimate": round(estimate, 3),
            "limit": params.limit,
            "window_s": window_s,
        }
        return allowed, state, None if allowed else round(window_s - elapsed, 3)

File: backend/app/limiters/scripts/sliding_counter.lua
-- Atomic sliding-window counter: weighted blend of current + previous fixed
-- windows. estimate = curr + prev * weight, where weight is the fraction of the
-- previous window still overlapping the trailing window.
-- KEYS[1] = sc:{client}:{window_index}        (current)
-- KEYS[2] = sc:{client}:{window_index - 1}    (previous)
-- ARGV[1] = limit
-- ARGV[2] = weight (0..1)
-- ARGV[3] = window TTL in ms (key kept for 2 windows)
-- Returns { allowed (0/1), curr_count, prev_count, estimate_string }
local limit  = tonumber(ARGV[1])
local weight = tonumber(ARGV[2])

local curr = tonumber(redis.call('GET', KEYS[1]) or '0')
local prev = tonumber(redis.call('GET', KEYS[2]) or '0')

local estimate = curr + prev * weight
local allowed = 0
if estimate < limit then
  curr = redis.call('INCR', KEYS[1])
  redis.call('PEXPIRE', KEYS[1], tonumber(ARGV[3]) * 2)
  allowed = 1
  estimate = curr + prev * weight
end

return { allowed, curr, prev, tostring(estimate) }
